### ISSUE TITLE
Firefox Extension: Can't unlock vault with PIN

### ISSUE THREAD

**Issue by DiGi** [bitwarden/clients#6286]
### Steps To Reproduce

1. Have Firefox extension logged in with enabled PIN unlock
2. Try to unlock with PIN
3. Enter key doesn't do anything, can't open vault

Extra steps
4. Logout from extension
5. Login back using main password and 2FA
6. Try to enable PIN again
7. "Enter PIN" field is not responding to Enter key, OK button is not working too

### Expected Result

I can unlock vault using PIN, I can enable Unlock by PIN option.

### Actual Result

Enter key is ignored, I can't submit existing PIN. I can't create new PIN.

### Additional Context

Extension is not working on my two desktop systems. 

I haven't any issues last week so it is maybe regression with new version 2023.8.3 (from Addons website: Last updated: 7 days ago (7. September 2023)).

There are may one or two star reviews from last week on Firefox Addons with login issues using biometrics too.

### Operating System

Windows

### Operating System Version

Windows 11 22H2, Windows 10 22H2

### Web Browser

Firefox

### Browser Version

117.0.1

### Build Version

Firefox Extension v2023.8.3

### Issue Tracking Info

- [X] I understand that work is tracked outside of Github. A PR will be linked to this issue should one be opened to address it, but Bitwarden doesn't use fields like "assigned", "milestone", or "project" to track progress.

**Comment by DiGi** [comment]
Duplicate of #6156 - I must messed "extension" while searching for open bug.

**Comment by DiGi** [comment]
This looks like new issue after releasing #6156.

**Comment by richardb64** [comment]
Is this a variation on part of the original problem, where in practice with a mismatched extension (#6156, #6185) it's actually very difficult to unlock the extension _at all_ - the biometric unlock dialogue is triggered (and then fails, due to the API breakage) without giving the user a chance to unlock in any other way.  The only workaround seems to be to hard-close the desktop client altogether, so that the biometric attempt never happens at all, and only then will the extension allow "local" login (master password and/or PIN).

That's probably worth raising as a separate bug in its own right (if there isn't already one) - ie. the extension should not plough into the biometric dialogue until the user actually clicks the "unlock with biometrics" option.

**Comment by ishanjain28** [comment]
The new release of bitwarden client is completely broken. 

I can not login with PIN anymore. I don't see any errors in console logs for the extension.

I tried to sign in(after logging out). It sends prelogin request and then crashes. For now, I have rolled back to the previous release.

**Comment by dinosmm** [comment]
I wonder if I am seeing a variant of the same issue, or if I should make a new issue?

When I enter the master password on the Firefox extension (2023.8.3), the extension window stays as is, no indication of being unlocked. If I click out of it, the extension icon still has a padlock on it, but if I click on it, the extension is unlocked and the vault accessible.
The padlock icon changes to a number icon if I refresh the page or (if already on the page) I click on a password entry to input it.

**Comment by MountComb** [comment]
Same issue on Firefox linux using version 2023.8.3.

The old version can be installed via <https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/versions/>, and version 2023.7.1 works fine. You may also need to turn off auto-update for this addon.

**Comment by stefan0xC** [comment]
Tried to reproduce this: If I downgrade the extension to 2023.7.1, setup the PIN, upgrade to 2023.8.3 it will lock the extension without asking for PIN (even though it will still say that I still have it enabled) and I also experience similar issues after unlocking with my password (i.e. the extension only syncs and changes the icon after I switch the active tab). Unchecking the "Unlock with PIN" option will let me unlock the extension with my password without any issues (besides not being able to setup Unlock with PIN).

**Comment by pwseo** [comment]
Seems to be a problem with only the version available via addons.mozilla.org.
If one loads the 2023.8.3 extension on the releases page from Github as a temporary extension, it works flawlessly.

Also, as someone pointed out on reddit, the extension on mozilla.org has only about 4 MB, while the one on Github releases page has about 8-9 MB; probably a defective (incomplete?) build.

**Comment by DiGi** [comment]
Related Reddit topic: https://www.reddit.com/r/Bitwarden/comments/16hulgu/unlock_with_pin_not_working_on_firefox_extension/

**Comment by Jastreb** [comment]
Reproduced 100% extension 2023.8.3 is broken on addons.mozilla.org please reupload

**Comment by pwseo** [comment]
I don't really know how many Bitwarden users use the Firefox extension, but it seems a bit odd that Bitwarden has not commented officially on this matter -- at least an email would be nice, telling people how to proceed in case they stumble upon this bug.
As it stands, people are left without access to their vault via the extension, and arguably without knowing what to do. Not everyone knows enough to visit reddit, or github, or even the Discourse forum.

**Comment by DiGi** [comment]
> I don't really know how many Bitwarden users use the Firefox extension

660k installs, but that's just total number. And not everyone use PIN or biometrics _on desktop_. But I completely agree that lack of information here on official source is annoying.

At least there is an "official" reply on Reddit:

> BitwardenJai (Bitwarden Employee) 3 days ago 
> Thanks to everyone both advising of this issue, and those who are actively sharing tips and workarounds! Our team is currently tracking this issue and researching how to best resolve.

**Comment by trmartin4** [comment]
Thank you for the report of this issue.  We are currently investigating and testing potential fixes for two (potentially related) issues around PIN usage on the Firefox extension:
- Inability to set a PIN, and
- Inability to log in with a PIN after it is set

We understand the frustration that this has caused and are working on a resolution as soon as possible.

**Comment by fooness** [comment]
> Seems to be a problem with only the version available via addons.mozilla.org.
> If one loads the 2023.8.3 extension on the releases page from Github as a temporary extension, it works flawlessly (besides disappearing on browser restart).

I can confirm that manually (and temporarily) installing the extension from GitHub — via https://github.com/bitwarden/clients/releases/tag/browser-v2023.8.3 — does (temporarily) fix the problem on Firefox `117.0.1` on macOS `13.5.2`.

**Comment by trmartin4** [comment]
@fooness, can you clarify that you were seeing the issue on macOS?  All reports thus far on this thread have (I believe) been on Windows clients.

**Comment by fooness** [comment]
@trmartin4 Yes, I can confirm that the issue happened to me on Firefox `117.0.1` on macOS `13.5.2`; sorry I did not even realize this issue is Windows-related.

**Comment by stefan0xC** [comment]
@trmartin4 It also happens on Linux (see https://github.com/bitwarden/clients/issues/6286#issuecomment-1720431307).

**Comment by pashynskykh** [comment]
Does not work for me on Firefox for both macOS and Windows. Had to roll back to the previous version of the extension.

**Comment by trmartin4** [comment]
@fooness did you also try re-installing the extension from addons.mozilla.org to see if the issue persisted?  We're trying to isolate whether it's fixed with re-installing the extension or specifically with the version on Github.

**Comment by pwseo** [comment]
@trmartin4 re-installing the extension from addons.mozilla.org does not work. I tried it on Windows 10 and on Linux (openSUSE Tumbleweed), both with the latest Firefox release.

Also, should we create an issue or escalate the current one to reflect the real problem? There seems to be a big difference between the extension provided via addons.mozilla.org and the one found on Github releases page. Who knows what else isn't working besides the PIN troubles mentioned up until now.

**Comment by fooness** [comment]
> @fooness did you also try re-installing the extension from addons.mozilla.org to see if the issue persisted? We're trying to isolate whether it's fixed with re-installing the extension or specifically with the version on Github.

For me it did not help; installing the extension from GitHub did fix the problem, though if I understood correctly, that locally installed extension won’t survive any browser restart. As others already noted, the extension via mozilla.org and github.com seem to somehow differ? Though maybe that’s only related to something something packaging.

**Comment by dinosmm** [comment]
>  testing potential fixes for two (potentially related) issues around PIN usage on the Firefox extension:
> 
> * Inability to set a PIN, and
> * Inability to log in with a PIN after it is set
> 

Are you aware there are more problems than just the PIN, e.g. my issue when I enter the master password that I described above?

**Comment by JLJTGR** [comment]
I *may* have a variant of this as well.

When using PIN unlock, viewing an item that is further protected by a Master Password prompt will fail with a JavaScript error(pressing OK does nothing visually).
Downgrading to 2023.7.1 fixes the issue.  My version of Firefox refuses to load the unsigned extension from GitHub regardless of `xpinstall.signatures.required` settings.
Removing the PIN and using only Master Passwords works for both prompts.

Although if I try the same thing in the Android app, it will CTD.  I sent a feedback about that through Google's crash prompts.

**Comment by pwseo** [comment]
@JLJTGR you have to load it as a temporary extension/addon, via the about:debugging page
Tweaking the `xpinstall.signatures.required` boolean value currently has no effect on the release channel.

**Comment by ParreyNZ** [comment]
I have this issue also. i have firefox 117.0.1 (64-bit) and with extension [v2023.8.3] loaded via addons.mozilla i cant set a pin. if i load the extension from github it works fine.
it also happens on my android phone firefox 117.0.1 and with extension [v2023.8.3] i cant set a pin using the extension. i cant load an older version of the extension on my phone as it wont let me.

**Comment by trmartin4** [comment]
> > testing potential fixes for two (potentially related) issues around PIN usage on the Firefox extension:
> > 
> > * Inability to set a PIN, and
> > * Inability to log in with a PIN after it is set
> 
> Are you aware there are more problems than just the PIN, e.g. my issue when I enter the master password that I described above? Would you be able to say if that is also being worked on?

@dinosmm This would be a different issue.  The investigation on this front has been limited to problems with PIN set up and unlock.  Could you please enter a separate Github issue?

**Comment by wavemaster447** [comment]
@trmartin4 I don't believe this is a separate issue - the most common symptom is inability to unlock your vault via PIN, but the root cause is the same - the 2023.8.3 build on [addons.mozilla.org ](https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/versions/)is less than half the size of all other builds,[ 3.99mb](https://i.imgur.com/6TKx5Kh.png) compared to 8-9MB for other builds. Github indicates 2023.8.3 should be 8.5MB.   Something is broken in the build process, or there was a botched upload to addons.mozilla.org.

**Comment by stefan0xC** [comment]
@dinosmm I think your issue might be related to this issue insofar that the code responsible for unlocking probably throws an exception because you have set up vaultwarden to be unlocked with PIN. If you remove this setting the issue should disappear. (It may be necessary to downgrade to v2023.7.1 to uncheck the Unlock with PIN option as the checkmark may not be active. Not sure what causes this but I could not find a way to specifically replicate that situation.)

@wavemaster447 as far as I've compared the two exentsions the main difference (besides the different mangling which point to a separate build process) seems to be the removal of the .map files which at least would account for the difference in file size of the firefox extension. (Not sure if the behavior is different as I'm not sure how to install the zip file from the release page.)

**Comment by wavemaster447** [comment]
Interesting, it looks like MAP files are removed from builds now as of #6224.

**Comment by trmartin4** [comment]
We are working on a fix for this issue and have a potential resolution that we are testing now.  If you are comfortable installing an unsigned extension and to ensure that we have consistent replication steps, could you please do the following with the latest release version (`2023.8.3`)?

1. Download [v2023.8.3](https://github.com/bitwarden/clients/releases/tag/browser-v2023.8.3) of the addon
2. Open Firefox Developer Edition
3. Complete the steps [here](https://extensionworkshop.com/documentation/publish/signing-and-distribution-overview/#signing-your-addons) to allow unsigned extensions
4. Navigate to about:addons
5. Click gear icon and select “Install Add-on From File”
6. Select .zip

If you are still able to replicate the problem on this version, you can test the new version by doing the same steps, except using the build from [here](https://github.com/bitwarden/clients/actions/runs/6226134503) instead of the `2023.8.3` version.

**Comment by pwseo** [comment]
@trmartin4 the one from [the v2023.8.3 link](https://github.com/bitwarden/clients/releases/tag/browser-v2023.8.3) is the one some people already tested as temporary addons in Firefox release (or Firefox Developer Edition): there seem to be no issues with that version (tested on Windows 10 and Linux).

**Comment by stefan0xC** [comment]
> The problem occurs with [the one from addons.mozilla.org](https://addons.mozilla.org/firefox/downloads/file/4164440/bitwarden_password_manager-2023.8.3.xpi) only.

@trmartin4 Just tested it and it also occurs with "[your suggested build](https://github.com/bitwarden/clients/actions/runs/6226134503)". I.e. I can't even set the "Unlock with PIN" option (and it asks for my master password even if I setup Unlock with PIN before installing this version).

Steps to try: download and unzip [the file](https://github.com/bitwarden/clients/suites/16294859687/artifacts/930432056) , then you can go to [`about:debugging` and select `This Firefox`](about:debugging#/runtime/this-firefox) and load the extracted `dist-firefox.zip` file as a temporary addon.

As far as I can tell the problem is that you are not building the extension from the released `v2023.8.3` tag (which works as confirmed by many users already) but from the `master` branch instead. Can you try to cherry-pick b63ffd5355334eab765e94ca5475e33fe0a5877a onto the `v2023.8.3` if you want us to test if the new building process is the issue or if there was a different change that caused the issue?

**Comment by Akaoni** [comment]
I had what people are citing as a possible "variant", insofar as the Password Re-Prompt was not responding to key or button press (I don't use PINs).

Removing the extension and re-adding fixed the issue (at least for now). I see people have downgraded, but has anyone else tried Remove/Add as a work-around?

Win Firefox 117.0.1; Bitwarden 2023.8.3

**Comment by trmartin4** [comment]
@Akaoni, if you experience any more issues with Password Re-Prompt, please let us know with another issue.  The problems here do appear to be tied to the PIN entry and not the re-prompt.

**Comment by trmartin4** [comment]
@stefan0xC thank you for that test. Just to confirm:

- ❌ Current extension on addons.mozilla.org 
- ❌ Extension loaded as temporary add-on from [suggested build](https://github.com/bitwarden/clients/actions/runs/6226134503)
- ✅ Extension loaded as temporary add-on from [`browser-v2023.8.3`](https://github.com/bitwarden/clients/releases/tag/browser-v2023.8.3)

To further test the theory proposed, I have cherry-picked https://github.com/bitwarden/clients/commit/b63ffd5355334eab765e94ca5475e33fe0a5877a into a branch created from the `browser-2023.8.3` tag, with the artifacts available [here](https://github.com/bitwarden/clients/actions/runs/6243321563).  If you are able to test this build it would provide valuable information for us as we track down this problem.  We truly appreciate the time you have spent helping the team narrow this down.

**Comment by stefan0xC** [comment]
@trmartin4 Yeah, that's right. I've just tested it and it does not work. After trying to replicate what happens locally I think I found the issue: By copying the files to a new directory you are forgetting to copy `patches/argon2-browser+1.18.0.patch`  to the `browser-source` directory.
https://github.com/bitwarden/clients/blob/284eda6cd30301ecba0bbb54ff0cbee45052a13a/.github/workflows/build-browser.yml#L139-L160

**Comment by imjuzcy** [comment]
A few suggestions:

- Change the title of this issue to remove the word "Windows", as it also happens to Firefox on Linux (I'm using Flatpak on Fedora, other users can confirm for other distributions/packaging)
- Pull the problematic version from addons.mozilla.org, since it's practically unusable. I think the vast majority of user have at least PIN or Master Password prompt and not without any lock at all
- Or better, post a rollback "update" that is the previous working version but with higher version number than the problematic one, so that users that are not aware of this thread can use the extension when Firefox eventually auto updates the extension

**Comment by pwseo** [comment]
@trmartin4, as @stefan0xC said, your 2023.9.0 version has the same issue (loaded it as a temporary extension).

Is it not time for some damage control? This will be the second version of the extension that does not work, and there has not been an official stance from Bitwarden regarding the issue.
Also, wouldn't it be more prudent to at least beta-test these before uploading them to their respective extension "stores"?

**Comment by pwseo** [comment]
@trmartin4 Just tested the extension from [the artifacts as you instructed @stefan0xC](https://github.com/bitwarden/clients/actions/runs/6250577828): it seems to be working.

**Comment by stefan0xC** [comment]
Can confirm that setting up Unlock with PIN, locking and unlocking with PIN  works with this version again.

**Comment by okolvik** [comment]
So when will this actually be pushed as an official update?

**Comment by trmartin4** [comment]
Version `2023.9.1` of the Firefox extension includes the fix to this bug.  It was submitted for review to all app stores last week and will begin rolling out as soon as it is approved.

**Comment by xxhh1112** [comment]
qewq

**Comment by pwseo** [comment]
> Version `2023.9.1` of the Firefox extension includes the fix to this bug. It was submitted for review to all app stores last week and will begin rolling out as soon as it is approved.

Just out of curiosity @trmartin4, how long does it usually take to approve?

**Comment by mftcodes** [comment]
I didn't want to pen a new ticket, so I'll post my experience here. It is master password related, NOT PIN related. 

I'm getting this problem on Firefox browsers on Linux Gnome desktop, MacOS, and Windows 11.
Specifically, this happens when firefox is restored to a previous session, and I do not get the password typed correctly the first time. Even if I get it right the first time, if I show the password while typing to double check, it also will not work. If I close the browser down, restart it, and then try, it works.

This behavior, to be clear, is the master password. I do not use a pin or biometrics.

Firefox V 118.0.1 on all OS's
Bitwarden Extension version 2023.8.3

**Comment by pwseo** [comment]
@mftcodes I would also suggest to try the newest version of the extension (via GitHub release page) to see if that behaviour persists. If it doesn't, then it's fixed :) v2023.8.3 had some problems, so it's best to skip it.

**Comment by okolvik** [comment]
Does it really take over a week to get the extension officially updated? 2023.8.3 is still the official version for firefox.

**Comment by pwseo** [comment]
> Does it really take over a week to get the extension officially updated? 2023.8.3 is still the official version for firefox.

It does, unfortunately, but this is not a Bitwarden problem. It's a Mozilla problem -- 2023.9.1 is already on the Chrome Web Store.

**Comment by okolvik** [comment]
> > Does it really take over a week to get the extension officially updated? 2023.8.3 is still the official version for firefox.
> 
> It does, unfortunately, but this is not a Bitwarden problem. It's a Mozilla problem -- 2023.9.1 is already on the Chrome Web Store.

No, i understand that it's a mozilla issue. It just seems ridiculous

**Comment by trmartin4** [comment]
We are currently waiting on the Mozilla approval process to complete for the `2023.9.1` release. We have submitted to the store approval, but we have not had that approval yet. We absolutely understand the frustration in the delay.

**Comment by trmartin4** [comment]
Version `2023.9.1` has been approved by Mozilla and is now available on addons.mozilla.org for update.

**Comment by JLJTGR** [comment]
> I _may_ have a variant of this as well.
> 
> When using PIN unlock, viewing an item that is further protected by a Master Password prompt will fail with a JavaScript error(pressing OK does nothing visually). Downgrading to 2023.7.1 fixes the issue. My version of Firefox refuses to load the unsigned extension from GitHub regardless of `xpinstall.signatures.required` settings. Removing the PIN and using only Master Passwords works for both prompts.
> 
> Although if I try the same thing in the Android app, it will CTD. I sent a feedback about that through Google's crash prompts.

2023.9.1 solves my variant as well.  Even the Android CTD I mentioned.

**Comment by ahnkle** [comment]
Even with 2023.9.1 I found that I could set a pin but then closing the extensions window and opening it again, the "Unlock with pin" checkbox was cleared.

I needed to **log out completely** and log in again before the "Unlock wth pin" option would remain active.

**Comment by mrlinuxfish** [comment]
I have had to set the pin on Bitwarden Firefox extension around 8-10 times since installing it. My Chrome setup at work has remembered the pin without forgetting it. (both windows 11)

IIRC, the extension on Firefox / Linux doesn't have this issue, but I don't currently have a Linux box setup to verify

Windows 10 / 11 (had the issue before upgrading os)
